Support for Intellij 2024.1

The 1.33 release should support Intellij 2024.1. If the minimum-supported version needs to increase that's fine but ideally 2023 should still be supported if possible. 

So far only EAP releases are available for testing so it's not yet known if the failing build will continue to be a problem. 

GitHub Action failed with:

```
Error: Exception in thread "main" java.lang.NoClassDefFoundError: kotlinx/collections/immutable/PersistentList
	at com.intellij.openapi.extensions.impl.ExtensionPointImpl.<clinit>(ExtensionPointImpl.kt:71)
	at com.intellij.openapi.extensions.impl.ExtensionsAreaImpl.doRegisterExtensionPoint(ExtensionsAreaImpl.kt:212)
	at com.intellij.openapi.extensions.impl.ExtensionsAreaImpl.registerExtensionPoint(ExtensionsAreaImpl.kt:186)
	at com.intellij.core.CoreApplicationEnvironment.registerExtensionPoint(CoreApplicationEnvironment.java:224)
	at com.intellij.core.CoreApplicationEnvironment.registerExtensionPoint(CoreApplicationEnvironment.java:214)
	at com.intellij.core.CoreApplicationEnvironment.registerApplicationExtensionPoint(CoreApplicationEnvironment.java:229)
	at com.intellij.core.CoreApplicationEnvironment.<init>(CoreApplicationEnvironment.java:93)
	at com.intellij.core.CoreApplicationEnvironment.<init>(CoreApplicationEnvironment.java:72)
	at org.intellij.grammar.LightPsi$MyParsing.<init>(LightPsi.java:152)
	at org.intellij.grammar.LightPsi.<clinit>(LightPsi.java:55)
	at org.intellij.grammar.Main.main(Main.java:38)
Caused by: java.lang.ClassNotFoundException: kotlinx.collections.immutable.PersistentList
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
```

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -129,7 +129,8 @@ intellij {
         buildPlugin { enabled = true }
         downloadRobotServerPlugin { enabled = true }
         jarSearchableOptions { enabled = true }
-        patchPluginXml { enabled = true }
+        patchPluginXml { enabled = false }
+        listProductsReleases { enabled = false }
         prepareSandbox { enabled = true }
         prepareTestingSandbox { enabled = true }
         prepareUiTestingSandbox { enabled = true }
@@ -237,10 +238,7 @@ val generateSyntaxLexer = tasks.register<GenerateLexerTask>("generateSyntaxLexer
     sourceFile.set(file("src/main/jflex/io/github/intellij/dlanguage/lexer/DLanguageLexer.flex"))
 
     // target directory for lexer
-    targetDir.set("gen/io/github/intellij/dlanguage/")
-
-    // target classname, target file will be targetDir/targetClass.java
-    targetClass.set("DlangLexer")
+    targetOutputDir.set(file("gen/io/github/intellij/dlanguage/"))
 }
 
 val generateDocSyntaxLexer = tasks.register<GenerateLexerTask>("generateDocSyntaxLexer") {
@@ -248,10 +246,7 @@ val generateDocSyntaxLexer = tasks.register<GenerateLexerTask>("generateDocSynta
     sourceFile.set(file("src/main/kotlin/io/github/intellij/dlanguage/features/documentation/DDocLexer.flex"))
 
     // target directory for lexer
-    targetDir.set("gen/io/github/intellij/dlanguage/features/documentation")
-
-    // target classname, target file will be targetDir/targetClass.java
-    targetClass.set("_DDocLexer")
+    targetOutputDir.set(file("gen/io/github/intellij/dlanguage/features/documentation/"))
 }
 
 val copyGenScript = tasks.register<Copy>("copyGenScript") {

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -9,7 +9,7 @@ sentry = "7+"
 
 # plugins
 gradleIntelliJPlugin = "1.10.1" # cannot update to latest as it breaks verifyPlugin
-grammarkit = "2022.3.1"
+grammarkit = "2022.3.2.2"
 kotlin = "1.8.22"
 
 [plugins]

--- a/sdlang/build.gradle.kts
+++ b/sdlang/build.gradle.kts
@@ -49,15 +49,12 @@ val generateSyntaxLexer = tasks.register<GenerateLexerTask>("generateSyntaxLexer
     sourceFile.set(file("src/main/kotlin/io/github/intellij/dlanguage/sdlang/lexer/SDLangLexer.flex"))
 
     // target directory for lexer
-    targetDir.set("gen/io/github/intellij/dlanguage/sdlang/lexer")
-
-    // target classname, target file will be targetDir/targetClass.java
-    targetClass.set("_SDLangLexer")
+    targetOutputDir.set(file("gen/io/github/intellij/dlanguage/sdlang/lexer"))
 }
 
 val generateSyntaxParser = tasks.register<GenerateParserTask>("generateSyntaxParser") {
     sourceFile.set(file("src/main/kotlin/io/github/intellij/dlanguage/sdlang/parser/SDLangParser.bnf"))
-    targetRoot.set("gen")
+    targetRootOutputDir.set(file("gen"))
     pathToParser.set("io/github/intellij/dlanguage/sdlang/parser/SDLangParser.java")
     pathToPsiRoot.set("io/github/intellij/dlanguage/sdlang/psi")
 }

--- a/src/test/java/io/github/intellij/dlanguage/codeinsight/dcd/DCDCompletionServerTest.java
+++ b/src/test/java/io/github/intellij/dlanguage/codeinsight/dcd/DCDCompletionServerTest.java
@@ -1,13 +1,52 @@
 package io.github.intellij.dlanguage.codeinsight.dcd;
 
 import com.intellij.execution.configurations.GeneralCommandLine;
+import com.intellij.openapi.application.Application;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.projectRoots.*;
+import com.intellij.openapi.roots.OrderRootType;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.testFramework.VfsTestUtil;
 import com.intellij.util.PathUtil;
+import io.github.intellij.dlanguage.DlangSdkType;
 import io.github.intellij.dlanguage.LightDlangTestCase;
 import org.junit.Test;
 
+import java.io.File;
+
 public class DCDCompletionServerTest extends LightDlangTestCase {
 
+    // this method will be called prior to the test being run and MockSdk is read only
+    @Override
+    public Sdk getProjectJDK() {
+        // Get a tmp folder to be the root of our "fake sdk"
+        var tmp = new File(FileUtil.getTempDirectory());
+        sdkRoot = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(tmp);
+        var phobos = VfsTestUtil.createDir(sdkRoot, "phobos");
+        var druntime = VfsTestUtil.createDir(sdkRoot, "druntime");
+
+        // Configure a sdk
+        var dlangSdk = DlangSdkType.getInstance();
+        var sdk = ProjectJdkTable.getInstance().createSdk("dmd", dlangSdk);
+        var sdkModificator = sdk.getSdkModificator();
+        sdkModificator.setVersionString("2");
+        sdkModificator.addRoot(phobos.getUrl(), OrderRootType.SOURCES);
+        sdkModificator.addRoot(druntime.getUrl(), OrderRootType.SOURCES);
+        // Ensure it is saved to be detected by the later code
+        Application application = ApplicationManager.getApplication();
+        Runnable runnable = sdkModificator::commitChanges;
+        if (application.isDispatchThread()) {
+            application.runWriteAction(runnable);
+        } else {
+            application.invokeAndWait(() -> application.runWriteAction(runnable));
+        }
+        return sdk;
+    }
+
     private DCDCompletionServer dcd;
+    private VirtualFile sdkRoot;
 
     @Override
     public void setUp() throws Exception {
@@ -21,8 +60,8 @@ public class DCDCompletionServerTest extends LightDlangTestCase {
 
         final String expected = String.format("dcd-server --ignoreConfig --logLevel all -I %s -I %s -I %s",
             PathUtil.toSystemDependentName("/src"),
-            PathUtil.toSystemDependentName("/phobos"),
-            PathUtil.toSystemDependentName("/druntime"));
+            PathUtil.toSystemDependentName(sdkRoot.getPath() + "/phobos"),
+            PathUtil.toSystemDependentName(sdkRoot.getPath() + "/druntime"));
 
         assertEquals(expected, cmd.getCommandLineString());
     }

--- a/src/test/kotlin/io/github/intellij/dlanguage/LightDlangTestCase.kt
+++ b/src/test/kotlin/io/github/intellij/dlanguage/LightDlangTestCase.kt
@@ -31,7 +31,7 @@ abstract class LightDlangTestCase : LightPlatform4TestCase() {
         roots.putValue(OrderRootType.SOURCES, MockDir("phobos"))
         roots.putValue(OrderRootType.SOURCES, MockDir("druntime"))
 
-        return MockSdk("dmd", "", "2", roots) { DlangSdkType.getInstance() }
+        return MockSdk("dmd", "", "2", roots, DlangSdkType.getInstance())
     }
 
     fun addFileToModuleSource(filename: String, content: String? = null): VirtualFile {

--- a/src/test/kotlin/io/github/intellij/dlanguage/LightDlangTestCase.kt
+++ b/src/test/kotlin/io/github/intellij/dlanguage/LightDlangTestCase.kt
@@ -1,10 +1,7 @@
 package io.github.intellij.dlanguage
 
 import com.intellij.mock.MockPsiManager
-import com.intellij.openapi.projectRoots.Sdk
-import com.intellij.openapi.projectRoots.impl.MockSdk
 import com.intellij.openapi.roots.ModuleRootManager
-import com.intellij.openapi.roots.OrderRootType
 import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.psi.FileViewProvider
 import com.intellij.psi.SingleRootFileViewProvider
@@ -12,27 +9,11 @@ import com.intellij.testFramework.LightPlatform4TestCase
 import com.intellij.testFramework.LightVirtualFile
 import io.github.intellij.dlanguage.psi.DlangFile
 import com.intellij.testFramework.VfsTestUtil
-import com.intellij.util.containers.MultiMap
 
 /**
  * Provides a base for unit testing that requires in-memory D source files
  */
 abstract class LightDlangTestCase : LightPlatform4TestCase() {
-
-    companion object {
-        class MockDir(val dirName: String) : LightVirtualFile(dirName) {
-            override fun isDirectory(): Boolean = true
-        }
-    }
-
-    // this method will be called prior to the test being run and MockSdk is read only
-    override fun getProjectJDK(): Sdk {
-        val roots = MultiMap<OrderRootType, VirtualFile>(2)
-        roots.putValue(OrderRootType.SOURCES, MockDir("phobos"))
-        roots.putValue(OrderRootType.SOURCES, MockDir("druntime"))
-
-        return MockSdk("dmd", "", "2", roots, DlangSdkType.getInstance())
-    }
 
     fun addFileToModuleSource(filename: String, content: String? = null): VirtualFile {
         val sourcesRoot = ModuleRootManager.getInstance(module).sourceRoots[0]

